benchmark/spmv/cuda_linops.hpp

```diff
@@ -0,0 +1,468 @@
+    friend class gko::EnableCreateMethod<CuspCsrEx>;
+
+public:
+    using csr = gko::matrix::Csr<ValueType, IndexType>;
+    using mat_data = gko::matrix_data<ValueType, IndexType>;
+
+    void read(const mat_data &data) override
+    {
+        csr_->read(data);
+        size_t buffer_size;
+        this->set_size(gko::dim<2>{csr_->get_size()});
+
+        const auto id = this->get_gpu_exec()->get_device_id();
+        gko::device_guard g{id};
+        // TODO: There is a problem here !
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseCsrmvEx_bufferSize(
+            this->get_gpu_exec()->get_cusparse_handle(), algmode_, trans_,
+            this->get_size()[0], this->get_size()[1],
+            csr_->get_num_stored_elements(), &scalars.get_const_data()[0],
+            CUDA_R_64F, this->get_descr(), csr_->get_const_values(), CUDA_R_64F,
+            csr_->get_const_row_ptrs(), csr_->get_const_col_idxs(), nullptr,
+            CUDA_R_64F, &scalars.get_const_data()[1], CUDA_R_64F, nullptr,
+            CUDA_R_64F, CUDA_R_64F, &buffer_size));
+        GKO_ASSERT_NO_CUDA_ERRORS(cudaMalloc(&buffer_, buffer_size));
+        set_buffer_ = true;
+    }
+
+    void apply_impl(const gko::LinOp *b, gko::LinOp *x) const override
+    {
+        auto dense_b = gko::as<gko::matrix::Dense<ValueType>>(b);
+        auto dense_x = gko::as<gko::matrix::Dense<ValueType>>(x);
+        auto db = dense_b->get_const_values();
+        auto dx = dense_x->get_values();
+
+        const auto id = this->get_gpu_exec()->get_device_id();
+        gko::device_guard g{id};
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseCsrmvEx(
+            this->get_gpu_exec()->get_cusparse_handle(), algmode_, trans_,
+            this->get_size()[0], this->get_size()[1],
+            csr_->get_num_stored_elements(), &scalars.get_const_data()[0],
+            CUDA_R_64F, this->get_descr(), csr_->get_const_values(), CUDA_R_64F,
+            csr_->get_const_row_ptrs(), csr_->get_const_col_idxs(), db,
+            CUDA_R_64F, &scalars.get_const_data()[1], CUDA_R_64F, dx,
+            CUDA_R_64F, CUDA_R_64F, buffer_));
+    }
+
+
+    gko::size_type get_num_stored_elements() const noexcept
+    {
+        return csr_->get_num_stored_elements();
+    }
+
+    CuspCsrEx(std::shared_ptr<const gko::Executor> exec,
+              const gko::dim<2> &size = gko::dim<2>{})
+        : gko::EnableLinOp<CuspCsrEx, CuspBase>(exec, size),
+          csr_(std::move(
+              csr::create(exec, std::make_shared<typename csr::classical>()))),
+          trans_(CUSPARSE_OPERATION_NON_TRANSPOSE),
+          set_buffer_(false)
+    {
+#ifdef ALLOWMP
+        algmode_ = CUSPARSE_ALG_MERGE_PATH;
+#endif
+    }
+
+    ~CuspCsrEx() override
+    {
+        const auto id = this->get_gpu_exec()->get_device_id();
+        gko::device_guard g{id};
+        if (set_buffer_) {
+            try {
+                GKO_ASSERT_NO_CUDA_ERRORS(cudaFree(buffer_));
+            } catch (std::exception &e) {
```

Comment:
I would prefer `const &`.
If you do not use `e` in the catch, I think you can even remove the variable name (like in the functions) to avoid potential `Parameter not used` warnings.